File: services/ledger-query-service/src/projector.test.ts
import {
  ACTIVITY_KEY,
  accountKey,
  applyEvent,
  dayOf,
  dayStatsKey,
  paymentIndexKey,
  paymentKey,
  statsKey,
  type PaymentEvent,
  type RedisLike,
} from './projector';

/** In-memory stand-in for Redis - no server needed to test the projection. */
function fakeRedis() {
  const hashes = new Map<string, Record<string, string>>();
  const lists = new Map<string, string[]>();
  const sets = new Map<string, Set<string>>();
  const zsets = new Map<string, Map<string, number>>();
  const expiries = new Map<string, number>();

  const client: RedisLike = {
    async hset(key, values) {
      const hash = hashes.get(key) ?? {};
      for (const [field, value] of Object.entries(values)) hash[field] = String(value);
      hashes.set(key, hash);
    },
    async hincrby(key, field, increment) {
      const hash = hashes.get(key) ?? {};
      hash[field] = String(Number(hash[field] ?? 0) + increment);
      hashes.set(key, hash);
    },
    async lpush(key, value) {
      lists.set(key, [value, ...(lists.get(key) ?? [])]);
    },
    async ltrim(key, start, stop) {
      lists.set(key, (lists.get(key) ?? []).slice(start, stop + 1));
    },
    async zadd(key, score, member) {
      const zset = zsets.get(key) ?? new Map<string, number>();
      zset.set(member, score);
      zsets.set(key, zset);
    },
    async zremrangebyrank(key, start, stop) {
      const zset = zsets.get(key);
      if (!zset) return;
      const ranked = [...zset].sort((a, b) => a[1] - b[1]);
      const size = ranked.length;
      const first = start < 0 ? Math.max(size + start, 0) : start;
      const last = stop < 0 ? size + stop : Math.min(stop, size - 1);
      for (let i = first; i <= last && i < size; i++) zset.delete(ranked[i][0]);
    },
    async expire(key, seconds) {
      expiries.set(key, seconds);
    },
    async sadd(key, member) {
      const set = sets.get(key) ?? new Set<string>();
      sets.set(key, set);
      if (set.has(member)) return 0;
      set.add(member);
      return 1;
    },
  };

  return {
    client,
    balance: (id: string) => Number(hashes.get(accountKey(id))?.balanceCents),
    hash: (key: string) => hashes.get(key),
    payment: (id: string) => hashes.get(paymentKey(id)),
    feed: (id: string) =>
      [...(zsets.get(paymentIndexKey(id)) ?? new Map())]
        .sort((a, b) => b[1] - a[1])
        .map(([member]) => member),
    activity: () => (lists.get(ACTIVITY_KEY) ?? []).map((raw) => JSON.parse(raw)),
    ttl: (key: string) => expiries.get(key),
  };
}

const AT = '2024-03-05T10:00:00.000Z';
let nextEventId = 0;
beforeEach(() => {
  nextEventId = 0;
});
const id = () => `evt-${++nextEventId}`;

const created = (account: string, balanceCents: number): PaymentEvent => ({
  eventId: id(),
  type: 'account.created',
  accountId: account,
  name: `acct-${account}`,
  balanceCents,
  occurredAt: AT,
});

type PaymentType = Exclude<PaymentEvent['type'], 'account.created'>;

const event = (
  type: PaymentType,
  overrides: Partial<{
    paymentId: string;
    amountCents: number;
    occurredAt: string;
    failureReason: string;
  }> = {},
): PaymentEvent =>
  ({
    eventId: id(),
    type,
    paymentId: overrides.paymentId ?? 'pay-1',
    fromAccountId: 'a',
    toAccountId: 'b',
    amountCents: overrides.amountCents ?? 2_500,
    note: 'lunch',
    occurredAt: overrides.occurredAt ?? AT,
    failureReason: overrides.failureReason ?? 'SETTLEMENT_FAILED_SIMULATED',
  }) as PaymentEvent;

async function seed(redis: RedisLike) {
  await applyEvent(redis, created('a', 10_000));
  await applyEvent(redis, created('b', 500));
}

describe('account.created', () => {
  it('writes the opening balance', async () => {
    const r = fakeRedis();
    await applyEvent(r.client, created('a', 10_000));
    expect(r.hash(accountKey('a'))).toEqual({
      id: 'a',
      name: 'acct-a',
      balanceCents: '10000',
    });
  });

  // With more than one partition there is no ordering guarantee between
  // events keyed by account and events keyed by payment. Every balance
  // mutation is therefore a delta, so any arrival order lands on the same
  // number - which is what makes partitioning safe here.
  it('lands on the same balance whichever order the events arrive in', async () => {
    const opening = created('a', 10_000);
    const spend = event('payment.initiated', { amountCents: 2_500 });

    const inOrder = fakeRedis();
    await applyEvent(inOrder.client, opening);
    await applyEvent(inOrder.client, spend);

    const reversed = fakeRedis();
    await applyEvent(reversed.client, spend);
    await applyEvent(reversed.client, opening);

    expect(inOrder.balance('a')).toBe(7_500);
    expect(reversed.balance('a')).toBe(7_500);
  });

  it('still records identity when the account event arrives last', async () => {
    const r = fakeRedis();
    await applyEvent(r.client, event('payment.initiated', { amountCents: 100 }));
    await applyEvent(r.client, created('a', 10_000));
    expect(r.hash(accountKey('a'))).toMatchObject({ id: 'a', name: 'acct-a' });
    expect(r.balance('a')).toBe(9_900);
  });
});

// The read model has to be correct at every step of the saga, not just at the
// end - a wallet showing the wrong balance for one second is a wrong wallet.
describe('the saga, projected step by step', () => {
  it('initiated debits the sender only - the money is in clearing', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated'));

    expect(r.balance('a')).toBe(7_500);
    expect(r.balance('b')).toBe(500); // receiver has NOT been paid yet
    expect(r.payment('pay-1')).toMatchObject({ status: 'PROCESSING' });
  });

  it('completed credits the receiver and finishes the payment', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated'));
    await applyEvent(r.client, event('payment.completed'));

    expect(r.balance('a')).toBe(7_500);
    expect(r.balance('b')).toBe(3_000);
    expect(r.payment('pay-1')).toMatchObject({ status: 'COMPLETED' });
  });

  it('stuck leaves balances alone - the money is still in clearing', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated'));
    await applyEvent(r.client, event('payment.stuck'));

    expect(r.balance('a')).toBe(7_500); // still debited
    expect(r.balance('b')).toBe(500); // still not paid
    expect(r.payment('pay-1')).toMatchObject({
      status: 'AWAITING_REFUND',
      failureReason: 'SETTLEMENT_FAILED_SIMULATED',
    });
  });

  it('refunded returns the sender to exactly where they started', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated'));
    await applyEvent(r.client, event('payment.stuck'));
    await applyEvent(r.client, event('payment.refunded'));

    expect(r.balance('a')).toBe(10_000);
    expect(r.balance('b')).toBe(500); // receiver never saw a cent
    expect(r.payment('pay-1')).toMatchObject({ status: 'REFUNDED' });
  });

  it('failed touches no balance at all', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.failed', { failureReason: 'INSUFFICIENT_FUNDS' }));

    expect(r.balance('a')).toBe(10_000);
    expect(r.balance('b')).toBe(500);
    expect(r.payment('pay-1')).toMatchObject({
      status: 'FAILED',
      failureReason: 'INSUFFICIENT_FUNDS',
    });
  });

  it('keeps one row per payment, not one per lifecycle event', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated'));
    await applyEvent(r.client, event('payment.completed'));

    expect(r.feed('a')).toEqual(['pay-1']);
    expect(r.feed('b')).toEqual(['pay-1']);
  });

  it('keeps a payment in place in the feed as its status changes', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated', { paymentId: 'old', occurredAt: '2024-03-05T09:00:00.000Z' }));
    await applyEvent(r.client, event('payment.initiated', { paymentId: 'new', occurredAt: '2024-03-05T11:00:00.000Z' }));
    // The older payment settles last - it must not jump to the top.
    await applyEvent(r.client, event('payment.completed', { paymentId: 'old', occurredAt: '2024-03-05T09:00:00.000Z' }));

    expect(r.feed('a')).toEqual(['new', 'old']);
  });

  it('records every lifecycle step in the global activity log', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated'));
    await applyEvent(r.client, event('payment.completed'));

    expect(r.activity().map((a) => a.type)).toEqual([
      'payment.completed',
      'payment.initiated',
    ]);
  });
});

describe('statistics', () => {
  it('counts money only once it has actually arrived', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated'));

    // In flight is not sent.
    expect(r.hash(statsKey('a'))).toBeUndefined();

    await applyEvent(r.client, event('payment.completed'));
    expect(r.hash(statsKey('a'))).toMatchObject({ sentCents: '2500', sentCount: '1' });
    expect(r.hash(statsKey('b'))).toMatchObject({ receivedCents: '2500', receivedCount: '1' });
  });

  it('does not count a payment that got stuck and refunded', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated'));
    await applyEvent(r.client, event('payment.stuck'));
    await applyEvent(r.client, event('payment.refunded'));

    expect(r.hash(statsKey('a'))).toBeUndefined();
    expect(r.hash(statsKey('b'))).toBeUndefined();
  });

  it('buckets by UTC day and expires the bucket', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.completed'));

    expect(dayOf(AT)).toBe('2024-03-05');
    expect(r.hash(dayStatsKey('a', '2024-03-05'))).toMatchObject({ sentCents: '2500' });
    expect(r.ttl(dayStatsKey('a', '2024-03-05'))).toBeGreaterThan(0);
    // Lifetime totals must never expire.
    expect(r.ttl(statsKey('a'))).toBeUndefined();
  });

  it('accumulates across payments and days', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.completed', { paymentId: 'p1', amountCents: 1_000 }));
    await applyEvent(r.client, event('payment.completed', { paymentId: 'p2', amountCents: 250, occurredAt: '2024-03-06T08:00:00.000Z' }));

    expect(r.hash(statsKey('a'))).toMatchObject({ sentCents: '1250', sentCount: '2' });
    expect(r.hash(dayStatsKey('a', '2024-03-05'))).toMatchObject({ sentCents: '1000' });
    expect(r.hash(dayStatsKey('a', '2024-03-06'))).toMatchObject({ sentCents: '250' });
  });
});

describe('feed limits', () => {
  it('keeps only the most recent 100 payments per account', async () => {
    const r = fakeRedis();
    await seed(r.client);
    for (let i = 0; i < 120; i++) {
      await applyEvent(
        r.client,
        event('payment.initiated', {
          paymentId: `p${String(i).padStart(3, '0')}`,
          occurredAt: new Date(Date.parse(AT) + i * 1000).toISOString(),
        }),
      );
    }
    const feed = r.feed('a');
    expect(feed).toHaveLength(100);
    expect(feed[0]).toBe('p119'); // newest kept
    expect(feed).not.toContain('p000'); // oldest dropped
  });

  it('caps the global activity log', async () => {
    const r = fakeRedis();
    await seed(r.client);
    for (let i = 0; i < 250; i++) {
      await applyEvent(r.client, event('payment.initiated', { paymentId: `p${i}` }));
    }
    expect(r.activity()).toHaveLength(200);
  });
});

// Delivery is at-least-once from both ends - the outbox can re-publish after
// a crash, and Kafka can redeliver on a rebalance. Applying twice would
// invent money, so these are the tests that matter most.
describe('at-least-once delivery', () => {
  it('ignores a redelivered event instead of double-counting', async () => {
    const r = fakeRedis();
    await seed(r.client);
    const initiated = event('payment.initiated');

    expect(await applyEvent(r.client, initiated)).toBe(true);
    expect(await applyEvent(r.client, initiated)).toBe(false);
    expect(await applyEvent(r.client, initiated)).toBe(false);

    expect(r.balance('a')).toBe(7_500);
    expect(r.activity()).toHaveLength(1);
  });

  it('does not double-count a redelivered completion, or its statistics', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated'));
    const completed = event('payment.completed');
    await applyEvent(r.client, completed);
    await applyEvent(r.client, completed);

    expect(r.balance('b')).toBe(3_000);
    expect(r.hash(statsKey('b'))).toMatchObject({ receivedCents: '2500', receivedCount: '1' });
  });

  it('treats two distinct events with identical payloads as two payments', async () => {
    const r = fakeRedis();
    await seed(r.client);
    await applyEvent(r.client, event('payment.initiated', { paymentId: 'x' }));
    await applyEvent(r.client, event('payment.initiated', { paymentId: 'y' }));

    expect(r.balance('a')).toBe(5_000);
    expect(r.feed('a')).toHaveLength(2);
  });

  it('replaying the whole stream is a no-op on an already-built read model', async () => {
    const stream: PaymentEvent[] = [
      created('a', 10_000),
      created('b', 500),
      event('payment.initiated'),
      event('payment.completed'),
    ];
    const r = fakeRedis();
    for (const e of stream) await applyEvent(r.client, e);
    const snapshot = [r.balance('a'), r.balance('b'), r.activity().length];

    for (const e of stream) await applyEvent(r.client, e);
    expect([r.balance('a'), r.balance('b'), r.activity().length]).toEqual(snapshot);
  });

  it('rebuilds identical state when replayed into an empty read model', async () => {
    const stream: PaymentEvent[] = [
      created('a', 10_000),
      created('b', 500),
      event('payment.initiated'),
      event('payment.stuck'),
      event('payment.refunded'),
    ];

    const original = fakeRedis();
    for (const e of stream) await applyEvent(original.client, e);
    const rebuilt = fakeRedis();
    for (const e of stream) await applyEvent(rebuilt.client, e);

    expect(rebuilt.balance('a')).toBe(original.balance('a'));
    expect(rebuilt.balance('b')).toBe(original.balance('b'));
    expect(rebuilt.payment('pay-1')).toEqual(original.payment('pay-1'));
    expect(rebuilt.activity()).toEqual(original.activity());
  });
});

File: services/ledger-query-service/src/test/java/com/ledgerflow/query/domain/ProjectorTest.java
package com.ledgerflow.query.domain;

import static com.ledgerflow.query.domain.Projector.accountKey;
import static com.ledgerflow.query.domain.Projector.applyEvent;
import static com.ledgerflow.query.domain.Projector.dayOf;
import static com.ledgerflow.query.domain.Projector.dayStatsKey;
import static com.ledgerflow.query.domain.Projector.statsKey;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.entry;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

class ProjectorTest {

  private static final String AT = "2024-03-05T10:00:00.000Z";

  private int nextEventId;

  @BeforeEach
  void resetEventIds() {
    nextEventId = 0;
  }

  private String id() {
    return "evt-" + (++nextEventId);
  }

  private PaymentEvent created(String account, long balanceCents) {
    return new PaymentEvent(
        id(),
        Projector.ACCOUNT_CREATED,
        AT,
        account,
        "acct-" + account,
        balanceCents,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null);
  }

  private PaymentEvent event(String type) {
    return event(type, "pay-1", 2_500, AT, "SETTLEMENT_FAILED_SIMULATED");
  }

  private PaymentEvent event(String type, String paymentId) {
    return event(type, paymentId, 2_500, AT, "SETTLEMENT_FAILED_SIMULATED");
  }

  private PaymentEvent event(String type, long amountCents) {
    return event(type, "pay-1", amountCents, AT, "SETTLEMENT_FAILED_SIMULATED");
  }

  private PaymentEvent event(
      String type, String paymentId, long amountCents, String occurredAt, String failureReason) {
    return new PaymentEvent(
        id(),
        type,
        occurredAt,
        null,
        null,
        null,
        paymentId,
        "a",
        "b",
        amountCents,
        "lunch",
        failureReason,
        null,
        null);
  }

  private void seed(FakeStore store) {
    applyEvent(store, created("a", 10_000));
    applyEvent(store, created("b", 500));
  }

  @Nested
  @DisplayName("account.created")
  class AccountCreated {

    @Test
    @DisplayName("writes the opening balance")
    void writesOpeningBalance() {
      FakeStore store = new FakeStore();
      applyEvent(store, created("a", 10_000));
      assertThat(store.hash(accountKey("a")))
          .containsOnly(entry("id", "a"), entry("name", "acct-a"), entry("balanceCents", "10000"));
    }

    // With more than one partition there is no ordering guarantee between
    // events keyed by account and events keyed by payment. Every balance
    // mutation is therefore a delta, so any arrival order lands on the same
    // number - which is what makes partitioning safe here.
    @Test
    @DisplayName("lands on the same balance whichever order the events arrive in")
    void orderIndependent() {
      PaymentEvent opening = created("a", 10_000);
      PaymentEvent spend = event("payment.initiated", 2_500);

      FakeStore inOrder = new FakeStore();
      applyEvent(inOrder, opening);
      applyEvent(inOrder, spend);

      FakeStore reversed = new FakeStore();
      applyEvent(reversed, spend);
      applyEvent(reversed, opening);

      assertThat(inOrder.balance("a")).isEqualTo(7_500);
      assertThat(reversed.balance("a")).isEqualTo(7_500);
    }

    @Test
    @DisplayName("still records identity when the account event arrives last")
    void identityWhenLate() {
      FakeStore store = new FakeStore();
      applyEvent(store, event("payment.initiated", 100));
      applyEvent(store, created("a", 10_000));
      assertThat(store.hash(accountKey("a")))
          .contains(entry("id", "a"), entry("name", "acct-a"));
      assertThat(store.balance("a")).isEqualTo(9_900);
    }
  }

  // The read model has to be correct at every step of the saga, not just at the
  // end - a wallet showing the wrong balance for one second is a wrong wallet.
  @Nested
  @DisplayName("the saga, projected step by step")
  class SagaStepByStep {

    @Test
    @DisplayName("initiated debits the sender only - the money is in clearing")
    void initiatedDebitsSender() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated"));

      assertThat(store.balance("a")).isEqualTo(7_500);
      assertThat(store.balance("b")).isEqualTo(500); // receiver has NOT been paid yet
      assertThat(store.payment("pay-1")).containsEntry("status", "PROCESSING");
    }

    @Test
    @DisplayName("completed credits the receiver and finishes the payment")
    void completedCreditsReceiver() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated"));
      applyEvent(store, event("payment.completed"));

      assertThat(store.balance("a")).isEqualTo(7_500);
      assertThat(store.balance("b")).isEqualTo(3_000);
      assertThat(store.payment("pay-1")).containsEntry("status", "COMPLETED");
    }

    @Test
    @DisplayName("stuck leaves balances alone - the money is still in clearing")
    void stuckLeavesBalances() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated"));
      applyEvent(store, event("payment.stuck"));

      assertThat(store.balance("a")).isEqualTo(7_500); // still debited
      assertThat(store.balance("b")).isEqualTo(500); // still not paid
      assertThat(store.payment("pay-1"))
          .containsEntry("status", "AWAITING_REFUND")
          .containsEntry("failureReason", "SETTLEMENT_FAILED_SIMULATED");
    }

    @Test
    @DisplayName("refunded returns the sender to exactly where they started")
    void refundedRestoresSender() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated"));
      applyEvent(store, event("payment.stuck"));
      applyEvent(store, event("payment.refunded"));

      assertThat(store.balance("a")).isEqualTo(10_000);
      assertThat(store.balance("b")).isEqualTo(500); // receiver never saw a cent
      assertThat(store.payment("pay-1")).containsEntry("status", "REFUNDED");
    }

    @Test
    @DisplayName("failed touches no balance at all")
    void failedTouchesNoBalance() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.failed", "pay-1", 2_500, AT, "INSUFFICIENT_FUNDS"));

      assertThat(store.balance("a")).isEqualTo(10_000);
      assertThat(store.balance("b")).isEqualTo(500);
      assertThat(store.payment("pay-1"))
          .containsEntry("status", "FAILED")
          .containsEntry("failureReason", "INSUFFICIENT_FUNDS");
    }

    @Test
    @DisplayName("keeps one row per payment, not one per lifecycle event")
    void oneRowPerPayment() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated"));
      applyEvent(store, event("payment.completed"));

      assertThat(store.feed("a")).containsExactly("pay-1");
      assertThat(store.feed("b")).containsExactly("pay-1");
    }

    @Test
    @DisplayName("keeps a payment in place in the feed as its status changes")
    void keepsFeedPosition() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(
          store, event("payment.initiated", "old", 2_500, "2024-03-05T09:00:00.000Z", null));
      applyEvent(
          store, event("payment.initiated", "new", 2_500, "2024-03-05T11:00:00.000Z", null));
      // The older payment settles last - it must not jump to the top.
      applyEvent(
          store, event("payment.completed", "old", 2_500, "2024-03-05T09:00:00.000Z", null));

      assertThat(store.feed("a")).containsExactly("new", "old");
    }

    @Test
    @DisplayName("records every lifecycle step in the global activity log")
    void recordsActivity() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated"));
      applyEvent(store, event("payment.completed"));

      List<Object> types = new ArrayList<>();
      for (Map<String, Object> entry : store.activity()) types.add(entry.get("type"));
      assertThat(types).containsExactly("payment.completed", "payment.initiated");
    }
  }

  @Nested
  @DisplayName("statistics")
  class Statistics {

    @Test
    @DisplayName("counts money only once it has actually arrived")
    void countsOnlyOnArrival() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated"));

      // In flight is not sent.
      assertThat(store.hash(statsKey("a"))).isNull();

      applyEvent(store, event("payment.completed"));
      assertThat(store.hash(statsKey("a")))
          .containsEntry("sentCents", "2500")
          .containsEntry("sentCount", "1");
      assertThat(store.hash(statsKey("b")))
          .containsEntry("receivedCents", "2500")
          .containsEntry("receivedCount", "1");
    }

    @Test
    @DisplayName("does not count a payment that got stuck and refunded")
    void doesNotCountRefunded() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated"));
      applyEvent(store, event("payment.stuck"));
      applyEvent(store, event("payment.refunded"));

      assertThat(store.hash(statsKey("a"))).isNull();
      assertThat(store.hash(statsKey("b"))).isNull();
    }

    @Test
    @DisplayName("buckets by UTC day and expires the bucket")
    void bucketsByDay() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.completed"));

      assertThat(dayOf(AT)).isEqualTo("2024-03-05");
      assertThat(store.hash(dayStatsKey("a", "2024-03-05")))
          .containsEntry("sentCents", "2500");
      assertThat(store.ttl(dayStatsKey("a", "2024-03-05"))).isGreaterThan(0);
      // Lifetime totals must never expire.
      assertThat(store.ttl(statsKey("a"))).isNull();
    }

    @Test
    @DisplayName("accumulates across payments and days")
    void accumulatesAcrossDays() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.completed", "p1", 1_000, AT, null));
      applyEvent(
          store, event("payment.completed", "p2", 250, "2024-03-06T08:00:00.000Z", null));

      assertThat(store.hash(statsKey("a")))
          .containsEntry("sentCents", "1250")
          .containsEntry("sentCount", "2");
      assertThat(store.hash(dayStatsKey("a", "2024-03-05"))).containsEntry("sentCents", "1000");
      assertThat(store.hash(dayStatsKey("a", "2024-03-06"))).containsEntry("sentCents", "250");
    }
  }

  @Nested
  @DisplayName("feed limits")
  class FeedLimits {

    @Test
    @DisplayName("keeps only the most recent 100 payments per account")
    void keepsMostRecentHundred() {
      FakeStore store = new FakeStore();
      seed(store);
      for (int i = 0; i < 120; i++) {
        applyEvent(
            store,
            event(
                "payment.initiated",
                "p" + String.format("%03d", i),
                2_500,
                Instant.parse(AT).plusSeconds(i).toString().replace("Z", ".000Z"),
                null));
      }
      List<String> feed = store.feed("a");
      assertThat(feed).hasSize(100);
      assertThat(feed.get(0)).isEqualTo("p119"); // newest kept
      assertThat(feed).doesNotContain("p000"); // oldest dropped
    }

    @Test
    @DisplayName("caps the global activity log")
    void capsActivityLog() {
      FakeStore store = new FakeStore();
      seed(store);
      for (int i = 0; i < 250; i++) {
        applyEvent(store, event("payment.initiated", "p" + i));
      }
      assertThat(store.activity()).hasSize(200);
    }
  }

  // Delivery is at-least-once from both ends - the outbox can re-publish after
  // a crash, and Kafka can redeliver on a rebalance. Applying twice would
  // invent money, so these are the tests that matter most.
  @Nested
  @DisplayName("at-least-once delivery")
  class AtLeastOnceDelivery {

    @Test
    @DisplayName("ignores a redelivered event instead of double-counting")
    void ignoresRedelivered() {
      FakeStore store = new FakeStore();
      seed(store);
      PaymentEvent initiated = event("payment.initiated");

      assertThat(applyEvent(store, initiated)).isTrue();
      assertThat(applyEvent(store, initiated)).isFalse();
      assertThat(applyEvent(store, initiated)).isFalse();

      assertThat(store.balance("a")).isEqualTo(7_500);
      assertThat(store.activity()).hasSize(1);
    }

    @Test
    @DisplayName("does not double-count a redelivered completion, or its statistics")
    void ignoresRedeliveredCompletion() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated"));
      PaymentEvent completed = event("payment.completed");
      applyEvent(store, completed);
      applyEvent(store, completed);

      assertThat(store.balance("b")).isEqualTo(3_000);
      assertThat(store.hash(statsKey("b")))
          .containsEntry("receivedCents", "2500")
          .containsEntry("receivedCount", "1");
    }

    @Test
    @DisplayName("treats two distinct events with identical payloads as two payments")
    void distinctEventsAreDistinct() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.initiated", "x"));
      applyEvent(store, event("payment.initiated", "y"));

      assertThat(store.balance("a")).isEqualTo(5_000);
      assertThat(store.feed("a")).hasSize(2);
    }

    @Test
    @DisplayName("replaying the whole stream is a no-op on an already-built read model")
    void replayIsNoOp() {
      List<PaymentEvent> stream =
          List.of(
              created("a", 10_000),
              created("b", 500),
              event("payment.initiated"),
              event("payment.completed"));
      FakeStore store = new FakeStore();
      for (PaymentEvent e : stream) applyEvent(store, e);
      List<Object> snapshot =
          List.of(store.balance("a"), store.balance("b"), store.activity().size());

      for (PaymentEvent e : stream) applyEvent(store, e);
      assertThat(List.of(store.balance("a"), store.balance("b"), store.activity().size()))
          .isEqualTo(snapshot);
    }

    @Test
    @DisplayName("rebuilds identical state when replayed into an empty read model")
    void rebuildsIdenticalState() {
      List<PaymentEvent> stream =
          List.of(
              created("a", 10_000),
              created("b", 500),
              event("payment.initiated"),
              event("payment.stuck"),
              event("payment.refunded"));

      FakeStore original = new FakeStore();
      for (PaymentEvent e : stream) applyEvent(original, e);
      FakeStore rebuilt = new FakeStore();
      for (PaymentEvent e : stream) applyEvent(rebuilt, e);

      assertThat(rebuilt.balance("a")).isEqualTo(original.balance("a"));
      assertThat(rebuilt.balance("b")).isEqualTo(original.balance("b"));
      assertThat(rebuilt.payment("pay-1")).isEqualTo(original.payment("pay-1"));
      assertThat(rebuilt.activity()).isEqualTo(original.activity());
    }
  }

  @Nested
  @DisplayName("the trust boundary")
  class TrustBoundary {

    @Test
    @DisplayName("knows which types it can project")
    void knowsKnownTypes() {
      assertThat(Projector.isKnownEventType("account.created")).isTrue();
      assertThat(Projector.isKnownEventType("payment.initiated")).isTrue();
      assertThat(Projector.isKnownEventType("payment.held")).isTrue();
      assertThat(Projector.isKnownEventType("payment.approved")).isTrue();
      assertThat(Projector.isKnownEventType("payment.settlement_retrying")).isTrue();
      assertThat(Projector.isKnownEventType("payment.completed")).isTrue();
      assertThat(Projector.isKnownEventType("payment.failed")).isTrue();
      assertThat(Projector.isKnownEventType("payment.stuck")).isTrue();
      assertThat(Projector.isKnownEventType("payment.refunded")).isTrue();
      // Published by the write side, and deliberately not projected.
      assertThat(Projector.isKnownEventType("reconciliation.drift_detected")).isFalse();
      assertThat(Projector.isKnownEventType("something.new")).isFalse();
    }

    @Test
    @DisplayName("reads a timestamp the same way whether screening or scoring it")
    void parsesTimestamps() {
      assertThat(Projector.parseTimestamp(AT)).isEqualTo(Instant.parse(AT).toEpochMilli());
      assertThat(Projector.parseTimestamp("2024-03-05T10:00:00Z")).isNotNull();
      assertThat(Projector.parseTimestamp("not a date")).isNull();
      assertThat(Projector.parseTimestamp(null)).isNull();
    }

    @Test
    @DisplayName("a held payment is authorised, so it debits the sender")
    void heldDebitsSender() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.held"));

      assertThat(store.balance("a")).isEqualTo(7_500);
      assertThat(store.payment("pay-1")).containsEntry("status", "HELD_FOR_REVIEW");
    }

    @Test
    @DisplayName("approving a held payment moves no money, only its status")
    void approvedMovesNoMoney() {
      FakeStore store = new FakeStore();
      seed(store);
      applyEvent(store, event("payment.held"));
      applyEvent(store, event("payment.approved"));

      assertThat(store.balance("a")).isEqualTo(7_500);
      assertThat(store.payment("pay-1")).containsEntry("status", "PROCESSING");
    }
  }
}
